### ISSUE TITLE
Allow Full State Dict with 2D FSDP + TP

### ISSUE THREAD

**Issue by mvpatel2000** [pytorch/pytorch#120722]
### 🐛 Describe the bug

Torch does not allow 2D FSDP + TP to get FULL_STATE_DICT. However, if I remove checks here:
https://github.com/pytorch/pytorch/blob/3f62b05d31d4b29d60874b05adc0e5aedbad3722/torch/distributed/fsdp/_state_dict_utils.py#L303-L310
and warnings here:
https://github.com/pytorch/pytorch/blob/3f62b05d31d4b29d60874b05adc0e5aedbad3722/torch/distributed/fsdp/_state_dict_utils.py#L811-L815
it seems to work fine.

Is there a reason we need these checks? Can we remove them given things seem to work fine?

### Versions

Nightly.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

**Comment by wz337** [comment]
Hey @mvpatel2000. Thanks for raising the issue. Initially we wanted to disable it because the full_state_dict is only full state dict on data parallel dimension. So to avoid the discrepancy in naming and semantics, we disable it to let users just use SHARDED_STATE_DICT in 2D case. 

Just to make sure that we understand your use case, so you do want to save the full FSDP state_dict (but still sharded on TP dimension), right? 

cc. @fegin

**Comment by mvpatel2000** [comment]
@wz337 hm... in our case it seems to gather everything (including TP dimension)... Why would it not do that?

**Comment by wz337** [comment]
> @wz337 hm... in our case it seems to gather everything (including TP dimension)... Why would it not do that?

Hmm. Interesting... We didn't enable the full_state_dict path and nor did we test it before. iirc, the full_state_dict path currently is not aware of 2D and DTensorExtension. 

Hmm. You are saying without making any change, simply removing the assert gives you the full_state_dict as you expect?

**Comment by wanchaol** [comment]
I think dcp's get_state_dict need to be able to get a full_state_dict if user want to? the full_state_dict option on the code pointer is from FSDP, so it might be FSDP only without aware of TP?

**Comment by mvpatel2000** [comment]
> Hmm. You are saying without making any change, simply removing the assert gives you the full_state_dict as you expect?
Yep!

We have some model which is FSDP + TPed. We then want to gather the model in order to package it up for easy sharing, e.g. into a Huggingface checkpoint. You can see how we do this here:
https://github.com/mosaicml/llm-foundry/blob/main/llmfoundry/callbacks/hf_checkpointer.py#L230

Currently, we do:
```
use_full_state_dict = not state.is_model_ddp and isinstance(state_dict_model, FSDP)
state_dict_context = fsdp_state_dict_type_context(
  original_model, state_dict_type='full') if use_full_state_dict else contextlib.nullcontext()

with state_dict_context:
  state_dict = state_dict_model.state_dict()
```

With the new API, we can do:
```
from torch.distributed.checkpoint.state_dict import StateDictOptions, get_model_state_dict

state_dict = get_model_state_dict(
  state_dict_model,
  options=StateDictOptions(
    full_state_dict=True,
    cpu_offload=True,
  ),
)
```
and this seems to work fine! We then create a new model which does not have TP applied, load in the model state dict, and export via Huggingface APIs

**Comment by wz337** [comment]
Oh ok. 

> > Hmm. You are saying without making any change, simply removing the assert gives you the full_state_dict as you expect?
> > Yep!
> 
> We have some model which is FSDP + TPed. We then want to gather the model in order to package it up for easy sharing, e.g. into a Huggingface checkpoint. You can see how we do this here: https://github.com/mosaicml/llm-foundry/blob/main/llmfoundry/callbacks/hf_checkpointer.py#L230
> 
> Currently, we do:
> 
> ```
> use_full_state_dict = not state.is_model_ddp and isinstance(state_dict_model, FSDP)
> state_dict_context = fsdp_state_dict_type_context(
>   original_model, state_dict_type='full') if use_full_state_dict else contextlib.nullcontext()
> 
> with state_dict_context:
>   state_dict = state_dict_model.state_dict()
> ```
> 
> With the new API, we can do:
> 
> ```
> from torch.distributed.checkpoint.state_dict import StateDictOptions, get_model_state_dict
> 
> state_dict = get_model_state_dict(
>   state_dict_model,
>   options=StateDictOptions(
>     full_state_dict=True,
>     cpu_offload=True,
>   ),
> )
> ```
> 
> and this seems to work fine! We then create a new model which does not have TP applied, load in the model state dict, and export via Huggingface APIs

Oh. Ok. Thanks for the code pointer. So to clarify, there are two layers here, FSDP state dict and DCP's distributed state dict (I am referring to `get_model_state_dict`). 

When we call the generic `get_model_state_dict` with full_state_dict, it calls into `_gather_state_dict`, which will all_gather the full tensor for you regardless of whether it is 1D or 2D. This is the reason that you are seeing the correct result. https://github.com/pytorch/pytorch/blob/e2ee87d48bd2a70a18920c800da9fd0522be619d/torch/distributed/_state_dict_utils.py#L113

If you use the context manager full_state_dict for FSDP and remove the `RuntimeError`, it would give you 1D DTensor state_dict. So in order to avoid confusion, we disable it in FSDP's full_state_dict originally.

As long as users are using `get_model_state_dict` to do state_dict retrieval, I think I am fine with removing the warning and RuntimeError. @fegin Wdyt?

**Comment by mvpatel2000** [comment]
Ah got it, thanks for the clarification! 

It'd be great if we could remove the error/warning then as suggested. It makes our life a LOT easier to easily export models.

**Comment by bigning** [comment]
@wz337  there is one more issue though. [here](https://github.com/pytorch/pytorch/blob/4cf6d1172b25a0a7b773b40cd2401b9a6dad0fc1/torch/distributed/checkpoint/state_dict.py#L399-L401) it only offloads to CPU **AFTER** getting **ALL** the state_dict. It easily gets OOM if there are huge DTensor in the state_dict. Can we run the following code each time we get the DTensor in state_dict ? I reported this issue [here](https://github.com/pytorch/pytorch/issues/121062). 
```
return _gather_state_dict(
      state_dict, cpu_offload=info.cpu_offload, ranks_only=ranks_only
)
```

We had to add the following hook to avoid OOM before calling `get_model_state_dict `

```
cpu_offload = True
def dtensor_to_tensor_hook(
    module: nn.Module,
    state_dict: Dict[str, Any],
    prefix: str,
    *args: Any,
) -> Dict[str, Any]:
    dtensor_fqns = []
    for fqn in state_dict.keys():
        tensor = state_dict[fqn]
        if isinstance(tensor, DTensor):
            dtensor_fqns.append(fqn)
            tensor = tensor.full_tensor()
            if dist.get_global_rank() == 0:
                if cpu_offload:
                    tensor = tensor.cpu()
                state_dict[fqn] = tensor
    if dist.get_global_rank() != 0:
        for fqn in dtensor_fqns:
            del state_dict[fqn]
    return state_dict

hooks = []
for name, module in state_dict_model.named_modules():
    if isinstance(module, FSDP):
        hooks.append(
            module._register_state_dict_hook(
                dtensor_to_tensor_hook))

```

**Comment by mvpatel2000** [comment]
Reopening given @bigning 's issue

@wz337 or @wanchaol perhaps you two have thoughts?

**Comment by mvpatel2000** [comment]
oh hm... I can't reopen, if someone from torch team could do that, it would be great :)

**Comment by fegin** [comment]
> _gather_state_dict

_gather_state_dict actually perform the rank0_only + cpu_offload tensor by tensor, [code pointer.](https://github.com/pytorch/pytorch/blob/main/torch/distributed/_state_dict_utils.py#L151). @bigning Can you clarify the issue?

**Comment by bigning** [comment]
@fegin that issue only happens if the state_dict has DTensor. you actually have a [TODO here ](https://github.com/pytorch/pytorch/blob/main/torch/distributed/_state_dict_utils.py#L149-L150)to support DTensor, lol.  One minimal reproduce is [this](https://github.com/pytorch/pytorch/issues/121062#issue-2164454983).

For DTensors in state_dict, it will go through [this path](https://github.com/pytorch/pytorch/blob/4cf6d1172b25a0a7b773b40cd2401b9a6dad0fc1/torch/distributed/checkpoint/state_dict.py#L399-L401) to gather and off_load_to_cpu

**Comment by bigning** [comment]
@fegin  follow-up on this, were you able to reproduce the memory issue  ? Can we fix it in 2.3.1 ?

**Comment by mvpatel2000** [comment]
Bumping, any plans to include? @fegin ?

**Comment by bigning** [comment]
Can someone take a look at this ? Or will fsdp2 solve this ?

**Comment by wz337** [comment]
> Can someone take a look at this ? Or will fsdp2 solve this ?

@bigning 
Yes. Using distributed state dict with FSDP2 + TP would solve this. You can refer to the unit test `test_fully_shard_tp_2d_set_full_state_dict ` here for usage. 
https://github.com/pytorch/pytorch/pull/135725/files
